File: 2023/1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

// step1 result with STEP2=false, step2 result with STEP2=true
#define STEP2 true

static const char* digits[] = {"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};

bool starts_with(const char *str, const char *prefix) {
    size_t lenPrefix = strlen(prefix);
    size_t lenStr = strlen(str);
    return lenStr < lenPrefix ? false : strncmp(prefix, str, lenPrefix) == 0;
}

int get_digit(int i, char* line){
    char c = line[i];
    if('1' <= c && c <= '9'){
        return c-'0';
    }else{
#if STEP2
        for(int j=0; j<sizeof(digits)/sizeof(digits[0]); j++){
            if(starts_with(line+i, digits[j])) return j+1;
        }
#endif
        return 0;
    }
}

int main(){
    char buffer[128];
    int sum = 0;
    while(fgets(buffer, sizeof(buffer), stdin)){
        int first=0, last=0;
        int size = strnlen(buffer, sizeof(buffer));
        buffer[--size] = '\0'; //remove the '\n' at the end
        for(int i=0; i<size && first==0; i++){
            if((first = get_digit(i, buffer)) != 0)
                break;
        }
        for(int i=size-1; i>=0 && last==0; i--){
            if((last = get_digit(i, buffer)) != 0)
                break;
        }
        sum += first*10+last;
    }
    printf("sum : %d",sum);
    return EXIT_SUCCESS;
}

File: 2023/1.zig
const std = @import("std");
const io = std.io;
const ArrayList = std.ArrayList;

// step1 result with STEP2=false, step2 result with STEP2=true
const STEP2 = true;

const digits = [_][]const u8{ "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };

fn match(haystack: []const u8, needle: []const u8) bool {
    var i: usize = 0;
    while (i < haystack.len and i < needle.len and haystack[i] == needle[i]) {
        i += 1;
    }
    return i == needle.len; //if we reached the end of the needle that means match
}

fn getDigit(i: usize, line: []const u8) ?u8 {
    if (line[i] >= '1' and line[i] <= '9') {
        return line[i] - '0';
    } else {
        if (!STEP2) return null;
        for (digits, 0..) |digit, j| {
            if (match(line[i..], digit)) {
                return @as(u8, @intCast(j + 1));
            }
        }
        return null;
    }
}

pub fn main() anyerror!void {
    var buffer: [128]u8 = undefined;
    const stdin = io.getStdIn().reader();
    var sum: u32 = 0;
    while (true) {
        const maybeLine = try stdin.readUntilDelimiterOrEof(&buffer, '\n');
        if (maybeLine == null) break;
        const line = maybeLine.?;

        var first: u8 = 0;
        var last: u8 = 0;

        var i: usize = 0;
        while (i < line.len and first == 0) : (i += 1) {
            if (getDigit(i, line)) |d| {
                first = d;
                break;
            }
        }

        i = line.len;
        while (i > 0 and last == 0) : (i -= 1) {
            if (getDigit(i - 1, line)) |d| {
                last = d;
                break;
            }
        }
        sum += first * 10 + last;
    }
    std.debug.print("sum : {}\n", .{sum});
}
